common/grpc/protobuf-jackson/src/main/java/org/curioswitch/common/protobuf/json/ProtoFieldInfo.java

```diff
@@ -0,0 +1,367 @@
+          "Trying to determine enum class of non-enum type: " + field);
+      Message msgWithEnumValue =
+          containingPrototype
+              .newBuilderForType()
+              .addRepeatedField(
+                  field,
+                  containingPrototype
+                      .newBuilderForType()
+                      .newBuilderForField(field)
+                      .setField(
+                          mapKeyField.descriptor(), mapKeyField.descriptor().getDefaultValue())
+                      .setField(
+                          mapValueField.descriptor(), mapValueField.descriptor().getDefaultValue())
+                      .build())
+              .build();
+      try {
+        return messageClass
+            .getDeclaredMethod(
+                getMapValueOrThrowMethodName(),
+                new ProtoFieldInfo(mapKeyField.descriptor(), containingPrototype).javaClass())
+            .invoke(msgWithEnumValue, mapKeyField.descriptor().getDefaultValue())
+            .getClass();
+      } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+        throw new IllegalStateException("Could not find or invoke map item getter.", e);
+      }
+    }
+
+    // Repeated field.
+    // Enums always have at least one value, so we can call getValues().get(0) without checking.
+    Message msgWithEnumValue =
+        containingPrototype
+            .newBuilderForType()
+            .addRepeatedField(
+                valueField().descriptor(),
+                valueField().descriptor().getEnumType().getValues().get(0))
+            .build();
+    try {
+      return ((List<?>) getEnumAsClassMethod().invoke(msgWithEnumValue)).get(0).getClass();
+    } catch (IllegalAccessException | InvocationTargetException e) {
+      throw new IllegalStateException("Could not invoke enum getter for determining type.", e);
+    }
+  }
+
+  /**
+   * Return the Java {@link Class} that corresponds to the value of this field. Generally used for
+   * method resolution and casting generics.
+   */
+  Class<?> javaClass() {
+    if (isMapField() && valueJavaType() == JavaType.MESSAGE) {
+      Message mapEntry = containingPrototype.newBuilderForType().newBuilderForField(field).build();
+      return mapEntry.getField(mapEntry.getDescriptorForType().findFieldByName("value")).getClass();
+    }
+    switch (valueJavaType()) {
+      case INT:
+        return int.class;
+      case LONG:
+        return long.class;
+      case FLOAT:
+        return float.class;
+      case DOUBLE:
+        return double.class;
+      case BOOLEAN:
+        return boolean.class;
+      case STRING:
+        return String.class;
+      case BYTE_STRING:
+        return ByteString.class;
+      case ENUM:
+        return int.class;
+      case MESSAGE:
+        return containingPrototype
+            .newBuilderForType()
+            .newBuilderForField(valueField().descriptor())
+            .build()
+            .getClass();
```

Comment:
I believe this is the cleanest way to get the class that handles repeated and map fields correctly - for example, trying to get the default value would generally return a type-erased container and have to go through similar annoyingness as `enumClass()`

Luckily it only happens at code-gen time.